### PR TITLE
initializers for batch_normalization

### PR DESCRIPTION
**Pull request by fukatani**
Initialize `batch_normalization` link's gamma and beta using `initializers`. #1243

### CODE REVIEW

File: chainer/links/normalization/batch_normalization.py [hunk 1]
```diff
@@ -59,14 +60,19 @@ class BatchNormalization(link.Link):
...
     def __init__(self, size, decay=0.9, eps=1e-5, dtype=numpy.float32,
-                 use_gamma=True, use_beta=True):
+                 use_gamma=True, use_beta=True,
+                 initial_gamma=None, initial_beta=None):
         super(BatchNormalization, self).__init__()
         if use_gamma:
             self.add_param('gamma', size, dtype=dtype)
-            self.gamma.data.fill(1)
+            if initial_gamma is None:
+                initial_gamma = numpy.ones(size)
```
Comment (delta2323) [inline chainer/links/normalization/batch_normalization.py]: Use `chainer.Initializers.One` instead of `ndarray` because `init_weight` implicitly converts `ndarray` to `Initializer` [here](https://github.com/pfnet/chainer/blob/master/chainer/initializers/__init__.py#L46).

File: chainer/links/normalization/batch_normalization.py [hunk 2]
```diff
@@ -59,14 +60,19 @@ class BatchNormalization(link.Link):
...
+            initializers.init_weight(self.gamma.data, initial_gamma)
         if use_beta:
             self.add_param('beta', size, dtype=dtype)
-            self.beta.data.fill(0)
+            if initial_beta is None:
+                initial_beta = numpy.zeros(size)
```
Comment (delta2323) [inline chainer/links/normalization/batch_normalization.py]: ditto (Use `chainer.Initializers.Zero`).

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 3]
```diff
@@ -243,4 +243,23 @@ def test_backward_gpu(self):
...
+        self.initial_gamma = numpy.random.uniform(-1, 1, self.size)
+        self.initial_gamma = self.initial_gamma.astype(numpy.float32)
+        self.initial_beta = numpy.random.uniform(-1, 1, self.size)
+        self.initial_beta = self.initial_beta.astype(numpy.float32)
+        self.link = links.BatchNormalization(self.size, self.decay,
+                                             initial_gamma=self.initial_gamma,
+                                             initial_beta=self.initial_beta)
+
+    @condition.retry(3)
+    def test_initialize(self):
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: Please check it works with both CPU and GPU.

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 4]
```diff
@@ -243,4 +243,23 @@ def test_backward_gpu(self):
...
+        self.assertTrue((self.initial_gamma == self.link.gamma.data).all())
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: Use `chainer.testing.assert_allclose` to compare arrays.

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 5]
```diff
@@ -243,4 +243,23 @@ def test_backward_gpu(self):
...
+        self.assertTrue((self.initial_beta == self.link.beta.data).all())
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: ditto (use `chainer.testing.assert_allclose`)

**Comment by delta2323** [comment]
Thank you for your PR. I added comments. So please check them.

**Comment by fukatani** [comment]
I appreciate your easy-to-understand comments. 
I reflected your pointed out, and add test for initialing without beta and gamma.

**Comment by delta2323** [comment]
@fukatani Thank you for the notification. I checked CI and also think that these failures are not related to this PR. I reran the test.

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 6]
```diff
@@ -258,8 +259,34 @@ def setUp(self):
                                              initial_beta=self.initial_beta)
 
     @condition.retry(3)
-    def test_initialize(self):
-        self.assertTrue((self.initial_gamma == self.link.gamma.data).all())
-        self.assertTrue((self.initial_beta == self.link.beta.data).all())
+    def test_initialize_cpu(self):
+        assert_allclose(self.initial_gamma, self.link.gamma.data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: Use `testing.assert_allclose` because OpenStack Style Guidelines, which we follow, forbids to import objects (see [here](http://docs.openstack.org/developer/hacking/#imports))

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 7]
```diff
@@ -258,8 +259,34 @@ def setUp(self):
...
+        assert_allclose(self.initial_beta, self.link.beta.data)
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_initialize_gpu(self):
+        self.link.to_gpu()
+        assert_allclose(self.initial_gamma, self.link.gamma.data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: ditto (use `testing.assert_allclose`)

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 8]
```diff
@@ -258,8 +259,34 @@ def setUp(self):
...
+
+class TestInitializeWithoutGammaAndBeta(unittest.TestCase):
+
+    def setUp(self):
+        self.decay = 0.9
+        self.size = 3
+        self.link = links.BatchNormalization(self.size, self.decay)
+
+    def test_initialize_cpu(self):
+        assert_allclose(numpy.ones(self.size), self.link.gamma.data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: ditto (use `testing.assert_allclose`)

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 9]
```diff
@@ -258,8 +259,34 @@ def setUp(self):
...
+        assert_allclose(numpy.zeros(self.size), self.link.beta.data)
+
+    @attr.gpu
+    def test_initialize_gpu(self):
+        self.link.to_gpu()
+        assert_allclose(numpy.ones(self.size), self.link.gamma.data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: ditto (use `testing.assert_allclose`)

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 10]
```diff
@@ -243,4 +244,49 @@ def test_backward_gpu(self):
...
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_initialize_gpu(self):
+        self.link.to_gpu()
+        assert_allclose(self.initial_gamma, self.link.gamma.data)
+        assert_allclose(self.initial_beta, self.link.beta.data)
+
+
+class TestInitializeWithoutGammaAndBeta(unittest.TestCase):
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: To me, the name of this class seems ambiguous because it also seems to check the behavior of `BatchNormalization` without `gamma` and `beta` attributes, although it is a matter of tastes. How about changing the name like `TestDefaultInitializer`?

**Comment by delta2323** [comment]
Thank you for applying my comments. I updated my comments. Please check them.

**Comment by delta2323** [comment]
LGTM 👍

**Comment by delta2323** [comment]
Thank you for your contribution!

**Comment by fukatani** [comment]
Thank you for merging!
It was a short PR, but like the idea of ​​naming, fruitful experience for me.
I'll try my best!
